The HardTimerAsEncoder.ino example got stuck on an ASSERT_FAULT.

lacklustrlabs opened the pull request:
I've tested this fix on a maple mini with 'compiler.warning_flags.all' selected.
The example is located here: Arduino_STM32/examples/Sensors/HardTimerAsEncoder/HardTimerAsEncoder.ino

lacklustrlabs commented:
I just want the example to work, maybe the example should be altered instead. It sends channel=0 and that's what causes the assert to trigger (with 'compiler.warning_flags.all' turned on).

Btw. I'm no big fan of the 'compiler.warning_flags.all' setting enabling extra debug code, there should be another setting for that. But that's another issue entirely.

victorpv commented:
Sorry haven't checked here in a while.
I still think would be better to ask Carlos why the encoder mode function takes a channel parameter. If it is not needed and no plan to need it in the future, then your change makes sense.
But that function has some commented code, perhaps he plans to add more functionality, in which case channel may be needed, and may be better to change the example.
If you can reach to Carlos thru the forum or github and find out, you can decide which change is more appropiate, the example, or the core.
I do wonder how did he get that example to work like it is, unless he made changes to the example of the core that he didn't push.

rogerclarkmelbourne commented:
see http://www.stm32duino.com/viewtopic.php?f=53&t=2283

lacklustrlabs commented:
Ok, I've reverted the changes in timer.c and made corrections in HardwareTimerAsEncoder.ino instead. 

The [changes](https://github.com/lacklustrlabs/Arduino_STM32/tree/timer_encoder_fix) does not show up here yet, maybe it's because the PR is closed?

There are still (unrelated) issues with the example though:
- It uses D4 & D5 as hardcoded input to Timer2, that can't always be true because MapleMini uses D10 & D11
- `timer.attachInterrupt(0, func)` must use the invalid channel 0, or there will be no overflow interrupts.
- General confusion about period and frequency in the .ino

rogerclarkmelbourne commented:
I reopened.

How does it look now ??

lacklustrlabs commented:
Thanks, the 'Files changes' tab looks ok now.

victorpv commented:
@lacklustrlabs what do you think on changing the D2, D3 etc referenced to PA0, PA1, etc, to make more universal?

lacklustrlabs commented:
It seems like timer2:TIMER_CH1 always is PA0, at least for all STMF1 boards. 
But is that always true?
Is there a way to ask `timer2.pinNumberForChannel(TIMER_CH1)` or something similar?

@victorpv  I'm currently testing your suggestion now, will push it in a moment

lacklustrlabs commented:
Now using PA0 and PA1 as timer input pin numbers.
Last checkin has some major cleanup as well.

There is a problem however, I can't make the frigging thing to count in the correct direction.
I've found some bugs in the `timer.setMode(TIMER_CH1, TIMER_ENCODER)` method. 
It should really set these register bits:
  
Pseudo code:
```
  timer.CCMR1.CC1S.set(01); // CC1 channel is configured as input, IC1 is mapped on TI1
  timer.CCMR1.CC2S.set(01); // CC2 channel is configured as input, IC2 is mapped on TI2.
  
  timer.CCER.CC1P.set(); // CC1P&CC1NP = 11 == noninverted, trigger on both edges
  timer.CCER.CC1NP.set(); 
  
  timer.CCER.CC2P.set(); // CC2P&CC2NP = 11 == noninverted, trigger on both edges
  timer.CCER.CC2NP.set();
  
  timer.CCMR1.IC1F.clear(); // No filter, sampling is done at fDTS
  timer.CCMR1.IC2F.clear(); // No filter, sampling is done at fDTS
```
But even when I fix those bits it still counts down on the ..,00,01,11,10,... sequence

lacklustrlabs commented:
In combination with #394 the example works on my Maple Minis

victorpv commented:
Thanks I'll test this in a couple other boards with an encoder.
Just a note, may be good to change all references to Dn pins to Pan, so it works in generic boards straight out of the box, but that's just the example anyway.

lacklustrlabs commented:
I'm thinking about switching all of the default pins to 5V tolerant pins, that would make it easier to interface real hardware. Thoughts?

victorpv commented:
Do you mean for the example sketch? if so, good idea.
I haven't had the time to test with an encoder yet :( hopefully soon.

stevstrong commented:
I started some reviews, any comment on those?

lacklustrlabs commented:
@stevstrong Hm, I can't see any review from you on this PR. I'll wait a bit, maybe it will show up.

Regardng this PR, I will do some tests to see if the changes I made to timer.c really are necessary, now that TIMER_CCMR1_CC2S_INPUT_TI2 will be corrected by your #323. I noticed that I write to one register that's only available in advanced registers.

lacklustrlabs commented on STM32F1/cores/maple/libmaple/timer.c:
```diff
@@ -329,16 +329,24 @@ static void output_compare_mode(timer_dev *dev, uint8 channel) {
-    (dev->regs).gen->CCMR1 = TIMER_CCMR1_CC1S_INPUT_TI1 | TIMER_CCMR1_CC2S_INPUT_TI2 | TIMER_CCMR1_IC2F | TIMER_CCMR1_IC1F ;
+    (dev->regs).gen->CCMR1 = TIMER_CCMR1_CC1S_INPUT_TI1 | TIMER_CCMR1_CC2S_INPUT_TI2;
+    // disable input capture filters
+    (dev->regs).gen->CCMR1 &= ~(TIMER_CCMR1_IC2F | TIMER_CCMR1_IC1F) ;
```
Yes, i know. It's there to make the intention clear. This could just as well be a comment

lacklustrlabs commented on STM32F1/libraries/A_STM32_Examples/examples/Sensors/HardTimerAsEncoder/HardTimerAsEncoder.ino:
```diff
@@ -26,92 +26,54 @@
+ */
 
 #include "HardwareTimer.h"
+#define SOUT Serial
```
If you want to use Serial3 instead of normal Serial, you can do that switch here in one single line, instead of having to find-and-replace the entire file. Come to think about it... maybe that's what Serial already is, - a macro for Serial1. I'll have to investigate that

lacklustrlabs replied:
Yes, Serial can be re-defined in an .ino accomplishing exactly the same thing. I'll remove SOUT.

lacklustrlabs commented on STM32F1/libraries/A_STM32_Examples/examples/Sensors/HardTimerAsEncoder/HardTimerAsEncoder.ino:
```diff
@@ -123,35 +85,78 @@ EENCODER SIMULATION PART.
+  //set mode, the channel is not used when in this mode.
+  // You must, however, use a valid channel number.
+  timer.setMode(TIMER_CH1, TIMER_ENCODER); 
+  timer.pause(); //stop... 
```
the timer.pause() is there from the original source. 
timer.setMode() starts the timer for some reason and I didn't change that.

lacklustrlabs commented on STM32F1/cores/maple/libmaple/timer.c:
```diff
@@ -215,7 +215,8 @@ void timer_disable(timer_dev *dev) {
  * @param mode New timer mode for channel
  */
 void timer_set_mode(timer_dev *dev, uint8 channel, timer_mode mode) {
-    ASSERT_FAULT(channel > 0 && channel <= 4);
```
This is the origin of this PR, this `ASSERT_FAULT(channel > 0 && channel <= 4); `code triggered when the old example sent in channel = 0.  
I changed it so that the assert would ignore channel when in encoder mode. But I did a rollback after you guys suggested i should correct the example instead.

ASSERT_FAULT triggers when condition is false, just like any ASSERT

STM32F1/system/libmaple/include/libmaple/util.h:
```
#define ASSERT_FAULT(exp)                       \
    if (exp) {                                  \
    } else {                                    \
        _fail(__FILE__, __LINE__, #exp);        \
    }
```

lacklustrlabs commented:
FML! timer(4) fails to trigger at overflow

victorpv commented:
The ISR is not being called at overflow?

lacklustrlabs commented:
@victorpv that's right, I'll take a look at this sometime later.

stevstrong commented on STM32F1/libraries/A_STM32_Examples/examples/Sensors/HardTimerAsEncoder/HardTimerAsEncoder.ino:
```diff
@@ -26,92 +26,53 @@
 
-unsigned long ints = 0;
+volatile unsigned long revolutions = 0;
 
```
I know that the original uses unsigned long, but I think an `uint32_t` would also satisfy the requirements.

lacklustrlabs replied:
yes, In embedded systems I prefer to be specific with the data types. I'll change to uint32_t, even though they are technically the same.

stevstrong commented on STM32F1/libraries/A_STM32_Examples/examples/Sensors/HardTimerAsEncoder/HardTimerAsEncoder.ino:
```diff
@@ -123,35 +84,78 @@ EENCODER SIMULATION PART.
+  timer.setOverflow(PPR);    //use this to match the number of pulse per revolution of the encoder. Most industrial use 1024 single channel steps. 
+  timer.setCount(0);          //reset the counter. 
+  timer.setEdgeCounting(TIMER_SMCR_SMS_ENCODER3); //or TIMER_SMCR_SMS_ENCODER1 or TIMER_SMCR_SMS_ENCODER2. This uses both channels to count and ascertain direction. 
+  timer.attachInterrupt(0, overflowInterrupt); //channel must be 0 here.  
```
Shouldn't be a `timer.refresh()` here, before `attachInterrupt()`?

lacklustrlabs replied:
YES! Thank you. that solved the ARR problem

lacklustrlabs commented:
Converted PR to a basic bug fix. It still requires #323 or #394.

The old commits are [here](https://github.com/lacklustrlabs/Arduino_STM32/tree/timer_encoder_fix_old). It tried to reduce global namespace pollution by using _static_ function variables and changed to timer1 (5V tolerant pins)

stevstrong commented:
I only have problems with the pin naming convention, sometimes is D2, D4, and other times is 5 and 4.
Even if in the current form is working on the maple mini, it would be better for other variants to be changed to PXY notation (e.g. PA2, PC14).

lacklustrlabs commented:
@stevstrong Yeah, the old branch fixed the pin naming too. But now I'm just trying to get the assert out of the way.

rogerclarkmelbourne commented:
I had problems when I tried to locally pull #323 .
If #394 contains one of the same fixes in #323, it will be easier for me to test and merge that one.

lacklustrlabs commented:
#394 is a small subset of #323, yes.
I tried to align the texts to minimize future merge conflicts.

rogerclarkmelbourne commented:
I pulled #394 an hour ago

rogerclarkmelbourne commented:
OK. So this is now just a change to one example

Sorry, but can you explain why then change is needed.

Does the timer channel not match the pin that was used ??

lacklustrlabs commented:
> Does the timer channel not match the pin that was used ??

Channel number is never used, except for triggering an assert.

The example called:
`timer.setMode(0, TIMER_ENCODER); //set mode, the channel is not used when in this mode. ` 

But STM32F1/cores/maple/libmaple/timer.c does an `ASSERT_FAULT(channel > 0 && channel <= 4)` test that halts the program if compiled with 'compiler.warning_flags.all'

My first commit to this PR altered the assert, but I later changed the example instead.
After that I made all kinds of cleanup of the example, but I realized that those changes made it harder to overview the core issue of this PR.

So here we are, back to basic.

I'll make a new PR sometime later with the pin numbering fix suggested by @stevstrong

stevstrong commented:
Compiles error free and looks reasonable.
So I merge this.
